File: testdata/fixtures/ruby/CWE-22.rb
# CWE-22 Path Traversal in Ruby
# File APIs that take a path built from user input allow directory traversal
# (../../etc/passwd) if not validated.

def show
  p = params[:file]
  name = params.get("name")
  ck = cookies[:path]
  qs = request.query_string

  # ruleid: ruby-path-traversal
  File.read(p)

  # ruleid: ruby-path-traversal
  File.open(name)

  # ruleid: ruby-path-traversal
  IO.read(ck)

  # ruleid: ruby-path-traversal
  File.read(qs)

  # ok: ruby-path-traversal
  File.read("/tmp/fixed.txt")

  # ok: ruby-path-traversal
  File.open("/var/data/file.txt", "r")

  # ok: ruby-path-traversal
  IO.read("/etc/hosts")
end

File: testdata/fixtures/rust/CWE-22.rs
// CWE-22: Path traversal via File::open with untrusted input
use std::fs::File;

fn vuln() {
    let p = std::env::var("PATH").unwrap();
    // ruleid: rust-path-traversal
    let _ = File::open(p);
    let q = std::env::args().next().unwrap();
    // ruleid: rust-path-traversal
    let _ = std::fs::File::open(q);
    let r = std::env::var("FILE").unwrap();
    let propagated = r;
    // ruleid: rust-path-traversal
    let _ = std::fs::read(propagated);
}

fn safe() {
    // ok: rust-path-traversal
    let _ = File::open("/tmp/fixed.txt");
    // ok: rust-path-traversal
    let _ = std::fs::read("/etc/hosts");
    // ok: rust-path-traversal
    let safe = "constant";
    let _ = std::fs::File::open(safe);
}
